### ISSUE TITLE
AssertionError when use flask+gunicorn+gevent

### ISSUE THREAD

**Issue by zhongxinghong** [joerick/pyinstrument#84]
Hai ! I encount this `AssertionError` when I use pyinstrument to hook my flask app like the [examples/flask_hello.py](https://github.com/joerick/pyinstrument/blob/master/examples/flask_hello.py) said.
```
AssertionError: Please raise an issue at http://github.com/pyinstrument/issues and let me know how you caused this error!
```
This error doesn't happen every time I send request. Instead, several consecutive requests occur once. Then Nginx will return 500 status code.

Here is the full trackback stack from gunicorn `error.log` file:
```
[2020-01-22 04:34:28 +0800] [26251] [ERROR] Error handling request /v2/services/xxxx/xxxx/get?key1=value1&key2=value2&key3=value3&__profile__
Traceback (most recent call last):
  File "/home/pyserver/.pyenv/versions/3.6.6/envs/envpy36/lib/python3.6/site-packages/gunicorn/workers/base_async.py", line 55, in handle
    self.handle_request(listener_name, req, client, addr)
  File "/home/pyserver/.pyenv/versions/3.6.6/envs/envpy36/lib/python3.6/site-packages/gunicorn/workers/ggevent.py", line 143, in handle_request
    super().handle_request(listener_name, req, sock, addr)
  File "/home/pyserver/.pyenv/versions/3.6.6/envs/envpy36/lib/python3.6/site-packages/gunicorn/workers/base_async.py", line 106, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
  File "/home/pyserver/.pyenv/versions/3.6.6/envs/envpy36/lib/python3.6/site-packages/flask/app.py", line 2463, in __call__
    return self.wsgi_app(environ, start_response)
  File "/home/pyserver/.pyenv/versions/3.6.6/envs/envpy36/lib/python3.6/site-packages/werkzeug/middleware/proxy_fix.py", line 228, in __call__
    return self.app(environ, start_response)
  File "/home/pyserver/.pyenv/versions/3.6.6/envs/envpy36/lib/python3.6/site-packages/flask/app.py", line 2449, in wsgi_app
    response = self.handle_exception(e)
  File "/home/pyserver/.pyenv/versions/3.6.6/envs/envpy36/lib/python3.6/site-packages/flask/app.py", line 1866, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/pyserver/.pyenv/versions/3.6.6/envs/envpy36/lib/python3.6/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/pyserver/.pyenv/versions/3.6.6/envs/envpy36/lib/python3.6/site-packages/flask/app.py", line 2446, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/pyserver/.pyenv/versions/3.6.6/envs/envpy36/lib/python3.6/site-packages/flask/app.py", line 1952, in full_dispatch_request
    return self.finalize_request(rv)
  File "/home/pyserver/.pyenv/versions/3.6.6/envs/envpy36/lib/python3.6/site-packages/flask/app.py", line 1969, in finalize_request
    response = self.process_response(response)
  File "/home/pyserver/.pyenv/versions/3.6.6/envs/envpy36/lib/python3.6/site-packages/flask/app.py", line 2266, in process_response
    response = handler(response)
  File "/home/pyserver/www/wsgi.py", line 44, in after_request
    output_html = g.profiler.output_html()
  File "</home/pyserver/.pyenv/versions/3.6.6/envs/envpy36/lib/python3.6/site-packages/pyinstrument/vendor/decorator.py:decorator-gen-9>", line 2, in output_html
  File "/home/pyserver/.pyenv/versions/3.6.6/envs/envpy36/lib/python3.6/site-packages/pyinstrument/util.py", line 37, in caller
    return func(*args, **kwargs)
  File "/home/pyserver/.pyenv/versions/3.6.6/envs/envpy36/lib/python3.6/site-packages/pyinstrument/profiler.py", line 101, in output_html
    return renderers.HTMLRenderer().render(self.last_session)
  File "/home/pyserver/.pyenv/versions/3.6.6/envs/envpy36/lib/python3.6/site-packages/pyinstrument/renderers/html.py", line 23, in render
    session_json = self.render_json(session)
  File "/home/pyserver/.pyenv/versions/3.6.6/envs/envpy36/lib/python3.6/site-packages/pyinstrument/renderers/html.py", line 69, in render_json
    return json_renderer.render(session)
  File "/home/pyserver/.pyenv/versions/3.6.6/envs/envpy36/lib/python3.6/site-packages/pyinstrument/renderers/jsonrenderer.py", line 36, in render
    frame = self.preprocess(session.root_frame())
  File "/home/pyserver/.pyenv/versions/3.6.6/envs/envpy36/lib/python3.6/site-packages/pyinstrument/session.py", line 77, in root_frame
    assert root_frame is None, ASSERTION_MESSAGE
AssertionError: Please raise an issue at http://github.com/pyinstrument/issues and let me know how you caused this error!
```
Here is my `gunicorn.conf.py` file
```python
import os
import multiprocessing

chdir = WSGI_APP_ROOT

daemon = True
accesslog = os.path.join(LOG_DIR, "access.log")
errorlog  = os.path.join(LOG_DIR, "error.log")

pidfile = os.path.join(RUN_DIR, "gunicorn.pid")

user  = 33  # www-data
group = 33  # www-data

bind = [ "unix:" + os.path.join(RUN_DIR, "gunicorn.sock") ]

workers = multiprocessing.cpu_count() * 2 + 1
worker_class = "gevent"
worker_connections = 1024
threads = multiprocessing.cpu_count() * 3 

timeout = 15
graceful_timeout = 15
keepalive = 5
```
Here is my `wsgi.py` module
```python
PYINSTRUMENT_HOOK          = True
PYINSTRUMENT_PROFILE_FIELD = "__profile__"

from gevent.monkey import patch_all
patch_all()

import os
from werkzeug.middleware.proxy_fix import ProxyFix
from v2.app import create_app

app = create_app("development")

if PYINSTRUMENT_HOOK:

    from flask import g, request, make_response
    from pyinstrument import Profiler

    @app.before_request
    def before_request():
        if PYINSTRUMENT_PROFILE_FIELD in request.args:
            g.profiler = Profiler()
            g.profiler.start()

    @app.after_request
    def after_request(response):
        if not hasattr(g, "profiler"):
            return response
        g.profiler.stop()
        output_html = g.profiler.output_html()
        return make_response(output_html)

app.wsgi_app = ProxyFix(app.wsgi_app, x_for=1)

```

The `requirements.txt` 
```
Flask==1.1.1
Werkzeug==0.15.4 
greenlet==0.4.15
gevent==1.4.0
gunicorn==20.0.4
pyinstrument==3.1.0
```
The pyenv version
```
Python 3.6.6
```

**Comment by joerick** [comment]
Thank you for the very detailed report! I'm guessing gevent's fiddling with the stack is throwing pyinstrument off. It might be that it's not possible to profile two concurrent requests with gevent, as there's only one setprofile receiver per-thread. But I'll try to find time to look into it, and at least get a better error message and prevent the crashing.

**Comment by shivdhar** [comment]
+1, same stack (flask+gunicorn+gevent).

I'd be happy to add a fix, if you could point me in the right direction @joerick!

**Comment by fzyzcjy** [comment]
Similar stack here (gevent + locust) for stress testing.

**Comment by vermaat** [comment]
Same here with eventlet

**Comment by joerick** [comment]
Hi all. So having thought about it the best solution might be to install a 'synthetic' root frame (called `Thread <num>`) at the top of the tree, and have all the gevent 'root' frames children of this frame.

If anyone wants to have a go at the above, please do, happy to look at a PR for it!

**Comment by moagstar** [comment]
@joerick we are also running into this issue. I would be happy to help fix it, but perhaps you could give some pointers on your proposed solution. Do you mean to insert a synthetic frame when capturing the call stack in ``Profiler`` or when we would encounter multiple root nodes in ``Session``?

**Comment by joerick** [comment]
Sorry for the delay in getting back @moagstar ! 

> Do you mean to insert a synthetic frame when capturing the call stack in `Profiler` or when we would encounter multiple root nodes in `Session`?

I think it would be better to record the thread of each snapshot in `Profile._call_stack_for_frame`, and then when building the frame tree in `Session.root_frame`, use that 'Thread X' frame as the root frame. I suppose we'd still have a 'single root frame assert', but now it would be asserting that the Profiler only captured frames on a single thread.

**Comment by joerick** [comment]
I think this should be fixed in v4.0.0. Let me know if not!

**Comment by Querela** [comment]
So, I still got the same error (only using `gevent` + `requests`). On version pyinstrument **4.1.1** with python 3.8, event 21.12.0.
If instead of `monkey.patch_all()` I disable patching for threads using `monkey.patch_all(thread=False)` it works fine.
